concept/type_/type_manager/validation/operation_time_validation.rs

```diff
@@ -70,32 +78,883 @@ macro_rules! object_type_match {
+        ) -> Result<(), SchemaValidationError> {
+            let type_having_instances = $existing_instances_validation_func(
+                snapshot,
+                type_manager,
+                thing_manager,
+                object_type.clone(),
+                interface_type.clone(),
+            )
+            .map_err(SchemaValidationError::ConceptRead)?;
+
+            if let Some(type_having_instances) = type_having_instances {
+                Err(SchemaValidationError::CannotOverrideCapabilityWithExistingInstances(
+                    $capability_kind,
+                    get_label_or_schema_err(snapshot, object_type)?,
+                    get_label_or_schema_err(snapshot, type_having_instances)?,
+                    get_label_or_schema_err(snapshot, interface_type)?,
+                ))
+            } else {
+                Ok(())
+            }
+        }
+    };
+}
+
+macro_rules! cannot_change_supertype_as_capability_with_existing_instances_is_lost_validation {
+    ($func_name:ident, $capability_kind:path, $object_type:ident, $get_lost_capabilities_func:path, $existing_instances_validation_func:path) => {
+        pub(crate) fn $func_name(
+            snapshot: &impl ReadableSnapshot,
+            type_manager: &TypeManager,
+            thing_manager: &ThingManager,
+            subtype: $object_type<'static>,
+            supertype: Option<$object_type<'static>>,
+        ) -> Result<(), SchemaValidationError> {
+            let lost_capabilities = $get_lost_capabilities_func(snapshot, subtype.clone(), supertype.clone())?;
+
+            for capability in lost_capabilities {
+                let interface_type = capability.interface();
+                let type_having_instances = $existing_instances_validation_func(
+                    snapshot,
+                    type_manager,
+                    thing_manager,
+                    subtype.clone(),
+                    interface_type.clone(),
+                )
+                .map_err(SchemaValidationError::ConceptRead)?;
+
+                if let Some(type_having_instances) = type_having_instances {
+                    return Err(SchemaValidationError::CannotChangeSupertypeAsCapabilityIsLostWhileHavingHasInstances(
+                        $capability_kind,
+                        get_label_or_schema_err(snapshot, subtype)?,
+                        get_opt_label_or_schema_err(snapshot, supertype)?,
+                        get_label_or_schema_err(snapshot, type_having_instances)?,
+                        get_label_or_schema_err(snapshot, interface_type)?,
+                    ));
+                }
+            }
+
+            Ok(())
+        }
+    };
+}
+
+macro_rules! capability_or_its_overriding_capability_with_violated_new_annotation_constraints {
+    ($func_name:ident, $capability_type:ident, $object_type:ident, $interface_type:ident, $existing_instances_validation_func:path) => {
+        paste! {
```

Comment:
I use paste here to hide these special "collection" functions for `Type`, `TypeAndSiblings`, `TypeAndSiblingsAndSubtypes` constraint validation modes. Otherwise I'd need to expose them as separate macros with their own implementations, while they should be only used in this context.